accepted/0000-caasp-nodes.md

```diff
@@ -0,0 +1,277 @@
+* Bootstrap/registration in the Uyuni/SUSE Manager context: registering a node to Uyuni/SUSE Manager via [the documented methods](https://opensource.suse.com/doc-susemanager/suse-manager/client-configuration/registration-overview.html).
+At the time of writing, registering a CaaS Platform node to Uyuni/SUSE Manager has to be manually performed by the user with:
+  * Bootstrap script
+  * UI: Systems > Bootstrapping
+  * `bootstrap` XMLRPC call
+  * Not usual but still possible and hackier: mass bootstrap via salt-ssh and a user-generated roster file
+
+## Assumptions
+
+We assume that:
+* The user is registering the CaaS Platform nodes as Salt clients
+* The registration of every CaaS Platform node to Uyuni/SUSE Manager is already completed by the user using the aforementioned methods
+* The user has already created an activation key and associated it with the onboarding. The CaaS Platform nodes are registered correctly and with the proper CaaS Platform channel(s) assigned.
+
+NOTE: It is outside of this RFC to discuss automatic registration methods of the CaaS Platform nodes.
+
+## Technical description of the problem
+
+### The forbidden actions
+
+[`skuba-update`](https://documentation.suse.com/suse-caasp/4/single-html/caasp-admin/#_base_os_updates) systemd timer will take care of [patching](https://github.com/SUSE/skuba/blob/master/skuba-update/skuba_update/skuba_update.py#L301-L305) all installed packages locally to every node of the cluster and optionally reboot the node if `zypper` signals that a reboot is required.
+
+Uyuni/SUSE Manager should not offer the possibility to reboot a registered CaaS Platform node.
+
+In general, patch and package upgrade/removal are safe operations for a CaaS Platform cluster: on the next trigger of `skuba-update`, if there are packages that required a reboot, `skuba-update` will orchestrate the reboot.
+
+Package upgrade/removal and patch apply is not a problem if issued from Uyuni/SUSE Manager, unless the involved packages of the action are one of the following:
+
+- `kubernetes-kubeadm`
+- `kubernetes-kubelet`
+- `kubernetes-client`
+- `cri-o`
+- `cni-plugins`
+
+NOTE: [installing NEW (not already installed) packages is allowed](https://documentation.suse.com/suse-caasp/4/single-html/caasp-admin/#_existing_cluster).
+
+At the time of the writing, Uyuni/SUSE Manager or Salt cannot forbid the package operations based on the target packages (there is no such kind of granularity).
+
+### Actions to forbid
+
+Decision criteria:
+
+* Either Uyuni/SUSE Manager allows patching and package upgrade/removal or it does not allow those operations.
+* Additionally, the list of the "forbidden" packages above may vary with time and with CaaS Platform upgrades. Currently, there is no programmatic way to gather a list of those packages from a CaaS Platform cluster. What if Uyuni/SUSE Manager hardcodes that list of packages and the list changes? An update of Uyuni/SUSE Manager would be required.
```

Comment:
> If you try to upgrade any of the above packages zypper will complain. This is described in depth here.

I manually created a situation where an updated version of `cri-o` is available:

```
ip-172-27-1-209:~ # zypper in --oldpackage cri-o-1.15.0-1.4
Loading repository data...
Reading installed packages...
Resolving package dependencies...

The following package is going to be downgraded:
  cri-o

1 package to downgrade.
Overall download size: 11.4 MiB. Already cached: 0 B. No additional space will be used or freed after the operation.
Continue? [y/n/v/...? shows all options] (y): y
Retrieving package cri-o-1.15.0-1.4.x86_64                                                                                                                                                                (1/1),  11.4 MiB ( 50.9 MiB unpacked)
Retrieving: cri-o-1.15.0-1.4.x86_64.rpm .................................................................................................................................................................................................[done]

Checking for file conflicts: ............................................................................................................................................................................................................[done]
(1/1) Installing: cri-o-1.15.0-1.4.x86_64 ...............................................................................................................................................................................................[done]
There are running programs which still use files and libraries deleted or updated by recent upgrades. They should be restarted to benefit from the latest updates. Run 'zypper ps -s' to list these programs.
 ```

and I upgraded it: `zypper` does not complain - is this because I manually created the setup?

```
ip-172-27-1-209:~ # rpm -qa | grep -iw cri-o
cri-o-1.15.0-1.4.x86_64
cri-o-kubeadm-criconfig-1.15.0-3.3.2.x86_64
ip-172-27-1-209:~ # rpm -qa | grep -i patterns-caasp
patterns-caasp-Node-1.15-1.2-1.1.x86_64
ip-172-27-1-209:~ # zypper up cri-o
Loading repository data...
Reading installed packages...
Resolving package dependencies...

The following package is going to be upgraded:
  cri-o

1 package to upgrade.
Overall download size: 11.3 MiB. Already cached: 0 B. No additional space will be used or freed after the operation.
Continue? [y/n/v/...? shows all options] (y): y
Retrieving package cri-o-1.15.0-3.3.2.x86_64                                                                                                                                                              (1/1),  11.3 MiB ( 50.9 MiB unpacked)
Retrieving: cri-o-1.15.0-3.3.2.x86_64.rpm ...................................................................................................................................................................................[done (3.3 MiB/s)]

Checking for file conflicts: ............................................................................................................................................................................................................[done]
(1/1) Installing: cri-o-1.15.0-3.3.2.x86_64 .............................................................................................................................................................................................[done]
There are running programs which still use files and libraries deleted or updated by recent upgrades. They should be restarted to benefit from the latest updates. Run 'zypper ps -s' to list these programs.
 ```

Reply:
> @ereslibre: can you imagine other packages SUMA should not touch?

I think `kernel` is one of those: if Uyuni/SUSE Manager install an upgrade or patch the kernel before `skuba-update`, the node does not have the reboot sentinel set. Have I understood it correctly?

Reply:
To be more specific: the dependency mechanism implemented by the `patterns-caasp-Node-<major>.<minor>` package:
  * **allows** patch release updates (like the one you similated before)
  * **prevents** major version updates
  * **minor** major version updates

Reply:
> I think `kernel` is one of those: if Uyuni/SUSE Manager install an upgrade or patch the kernel before `skuba-update`, the node does not have the reboot sentinel set. Have I understood it correctly?

Yes, it wouldn't have the reboot sentinel. That why I suggested [above](https://github.com/uyuni-project/uyuni-rfc/pull/26#discussion_r345285894) that you should `touch /var/run/reboot-required` when needed